### PR TITLE
Finalizing Soft Delete

### PR DESCRIPTION
**Pull request by philiptyler**
Added some key ideas I picked up from other soft delete projects and sitting down with some other TBT devs.

### CODE REVIEW

**Comment by philiptyler** [comment]
@dlo this is ready for merge/dist whenever you get a moment
I was going to do it, but I realize I don't have PyPi permissions to do so

**Review by dlo — CHANGES_REQUESTED** [review]

**Comment by dlo** [comment]
@philiptyler Didn't realize my comments didn't come through until I "added" the review. Oops! :)

File: lionheart/models.py [hunk 1]
```diff
@@ -200,34 +200,67 @@ class SoftDeleteQuerySet(models.query.QuerySet):
     SubClass of the standard Django QuerySet that ignores soft-deleted rows,
     and overwrites the delete function to update with soft-delete instead
     """
-    def delete(self):
-        return super(SoftDeleteQuerySet, self).update(deleted=SoftDeleteMixin.DELETED)
+    def delete(self, using='default', *args, **kwargs):
+        if not len(self):
```
Comment (philiptyler) [inline lionheart/models.py]: agreed. defensive coding lol

File: lionheart/models.py [hunk 2]
```diff
@@ -200,34 +200,67 @@ class SoftDeleteQuerySet(models.query.QuerySet):
...
+        for model in self:
+            model.delete(using, *args, **kwargs)
 
     def remove_permanently(self):
         return super(SoftDeleteQuerySet, self).delete()
 
     def deleted(self):
-        return super(SoftDeleteQuerySet, self).filter(deleted=SoftDeleteMixin.DELETED)
+        qs = super(SoftDeleteQuerySet, self).filter(deleted=SoftDeleteMixin.DELETED)
+        qs.__class__ = SoftDeleteQuerySet
```
Comment (philiptyler) [inline lionheart/models.py]: super(SoftDeleteQuerySet, self).filter will return a django.models.QuerySet, so any chained calls wont use SoftDeleteQuerySet Logic.  By setting the class to SoftDeleteQuerySet, chained calls will use the correct overridden functions

Comment (philiptyler) [inline lionheart/models.py]: Essentially we're just casting.
http://stackoverflow.com/questions/9165756/object-type-casting-in-python-after-reloading-a-module-for-on-the-fly-code-cha

Comment (dlo) [inline lionheart/models.py]: Got it, makes sense. The superclass filter method returns a plain old QuerySet. A comment here is probably worthwhile.

File: lionheart/models.py [hunk 3]
```diff
@@ -241,19 +239,19 @@ def remove_permanently(self):
...
     def filter(self, *args, **kwargs):
-        if 'pk' in kwargs or 'id' in kwargs:
-            qs = self.all_with_deleted().filter(*args, **kwargs)
-        else:
-            qs = self.get_queryset().filter(*args, **kwargs)
+        qs = self._queryset_from_kwarg_contidtions(kwargs).filter(*args, **kwargs)
         qs.__class__ = SoftDeleteQuerySet
         return qs
 
+    def _queryset_from_kwarg_contidtions(self, kwargs):
```
Comment (dlo) [inline lionheart/models.py]: Minor typo :)

**Review by dlo — APPROVED** [review]

**Comment by dlo** [comment]
Looks great! 🎉

**Pull request merged**